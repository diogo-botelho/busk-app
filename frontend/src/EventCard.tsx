import { Card } from "react-bootstrap";
import { Event } from "./interfaces/Event";

interface EventCardParams {
    event: Event
}

export function EventListCard({ event } : EventCardParams) {
  const { title, type } = event;

  return (
    <Card className="mb-3">
      <Card.Body>
        <Card.Title>{title}</Card.Title>
        <Card.Text>{type}</Card.Text>
      </Card.Body>
    </Card>
  );
}

// export default EventListCard;
