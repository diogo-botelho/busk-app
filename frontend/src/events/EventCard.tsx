import { useContext, useState } from "react";
import { Container, Card, Button } from "react-bootstrap";

import { Event } from "../interfaces/Event";
import { UpdateEventFormData } from "../interfaces/UpdateEventFormData";

import { UpdateEventForm } from "./UpdateEventForm";

import { UserContext } from "../users/UserContext";

interface EventCardParamsInterface {
  event: Event;
  updateEvent: (event: Event, formData: UpdateEventFormData) => void;
  removeEvent: (eventId: number) => void;
  toggleDynamicMarker: (show: boolean) => void;
  isAddingEvent: boolean;
}

/** Show information about an event.
 * 
 * Is rendered by EventList to show a "card" for each event. Includes buttons
 * to update or remove the event.
 *
 * Props:
 *  - event: event object
 *  - updateEvent(): function that handles updating event details.
 *  - removeEvent(): function that handles removing event.
 *  - toggleDynamicMarker(): function that handles showing/hiding DynamicMarker
 *  on Map component.
 *  - isAddingEvent: tracks if an event is being added and hides update/delete 
 *  buttons if yes.
 * 
 * Context:
 *  - currentUser: current logged in user, or undefined.
 * 
 * State:
 *  - isUpdatingEvent: tracks if event is being updated.
 *
 * EventList -> EventCard -> UpdateEventForm
 */

export function EventCard({
  event,
  updateEvent,
  removeEvent,
  toggleDynamicMarker,
  isAddingEvent,
}: EventCardParamsInterface) {
  const currentUser = useContext(UserContext);
  const [isUpdatingEvent, setIsUpdatingEvent] = useState(false);

  const { title, type, id, buskerId } = event;

  //Toggle Add Event Button
  function toggleUpdateEvent() {
    setIsUpdatingEvent(true);
    toggleDynamicMarker(true);
  }

  async function handleCancel() {
    setIsUpdatingEvent(false);
    toggleDynamicMarker(false);
  }

  async function handleRemove() {
    toggleDynamicMarker(false);
    await removeEvent(id);
  }

  return (
    <Card className="mb-3">
      {isUpdatingEvent ? (
        <Card.Body>
          <UpdateEventForm event={event} updateEvent={updateEvent} />
          <Button onClick={handleCancel}>Cancel</Button>
        </Card.Body>
      ) : (
        <Card.Body>
          {" "}
          <Card.Title>{title}</Card.Title>
          <Card.Text>{type}</Card.Text>
          {!isAddingEvent && buskerId === currentUser?.buskerId ? (
            <Container>
              <Button onClick={toggleUpdateEvent}>Update</Button>
              <Button onClick={handleRemove}>Remove</Button>
            </Container>
          ) : null}
        </Card.Body>
      )}
    </Card>
  );
}
