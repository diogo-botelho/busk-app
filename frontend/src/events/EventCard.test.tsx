import React from "react";
import { render } from "@testing-library/react";
import { EventCard } from "./EventCard";

const demoEvent = {
  id: 1,
  buskerId: 1,
  buskerName: "demoBusker",
  title: "demo title",
  type: "demo type",
  coordinates: { lat: 1, lng: 1 },
};

it("renders without crashing", function () {
  render(<EventCard event={demoEvent} />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<EventCard event={demoEvent} />);
  expect(asFragment()).toMatchSnapshot();
});
