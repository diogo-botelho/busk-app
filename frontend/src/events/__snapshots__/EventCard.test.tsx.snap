// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="mb-3 card"
  >
    <div
      class="card-body"
    >
      <div
        class="card-title h5"
      >
        demo title
      </div>
      <p
        class="card-text"
      >
        demo type
      </p>
    </div>
  </div>
</DocumentFragment>
`;
